synapse/http/proxyagent.py

```diff
@@ -263,93 +270,79 @@ def _http_proxy_endpoint(
+            This currently supports http:// and https:// proxies.
             A hostname without scheme is assumed to be http.
 
         reactor: reactor to be used to connect to the proxy
 
         tls_options_factory: the TLS options to use when connecting through a https proxy
+
         kwargs: other args to be passed to HostnameEndpoint
 
     Returns:
-        interfaces.IStreamClientEndpoint|None: endpoint to use to connect to the proxy,
-            or None
-
-    Raises: ValueError if given a proxy with a scheme we don't support.
+        a tuple of
+            endpoint to use to connect to the proxy, or None
+            ProxyCredentials or if no credentials were found, or None
     """
     if proxy is None:
-        return None
-
-    # Note: we can't use urlsplit/urlparse as that is completely broken for things without a scheme://
-    scheme, host, port = parse_proxy(proxy)
+        return None, None
 
-    if scheme not in (b"http", b"https"):
-        raise ValueError("Proxy scheme '{}' not supported".format(scheme.decode()))
+    # Note: urlsplit/urlparse cannot be used here as that does not work (for Python
+    # 3.9+) on scheme-less proxies, e.g. host:port.
+    scheme, host, port, credentials = parse_proxy(proxy)
 
     proxy_endpoint = HostnameEndpoint(reactor, host, port, **kwargs)
 
     if scheme == b"https":
         tls_options = tls_options_factory.creatorForNetloc(host, port)
         proxy_endpoint = wrapClientTLS(tls_options, proxy_endpoint)
 
-    return proxy_endpoint
-
-
-def parse_username_password(proxy: bytes) -> Tuple[Optional[ProxyCredentials], bytes]:
-    """
-    Parses the username and password from a proxy declaration e.g
-    username:password@hostname:port.
-
-    Args:
-        proxy: The proxy connection string.
-
-    Returns
-        An instance of ProxyCredentials and the proxy connection string with any credentials
-        stripped, i.e u:p@host:port -> host:port. If no credentials were found, the
-        ProxyCredentials instance is replaced with None.
-    """
-    if proxy and b"@" in proxy:
-        # We use rsplit here as the password could contain an @ character
-        credentials, proxy_without_credentials = proxy.rsplit(b"@", 1)
-        return ProxyCredentials(credentials), proxy_without_credentials
-
-    return None, proxy
+    return proxy_endpoint, credentials
 
 
 def parse_proxy(
     proxy: bytes, default_scheme: bytes = b"http", default_port: int = 1080
-) -> Tuple[bytes, bytes, int]:
+) -> Tuple[bytes, bytes, int, Optional[ProxyCredentials]]:
     """
-    Parse the scheme, hostname and port from a proxy connection byte string.
+    Given a HTTP proxy URL, breaks it down into components and checks that it
+    has a hostname (otherwise it is not right useful to us trying to find a
+    proxy) and asserts that the URL has the scheme as that is all we support.
+    Parse the scheme, username, password, hostname and port from a proxy connection byte string.
```

Comment:
docstrings should start with a one-line summary:

```suggestion
    Parse a proxy connection string.
    
    Given a HTTP proxy URL, breaks it down into components and checks that it
    has a hostname (otherwise it is not useful to us when trying to find a
    proxy) and asserts that the URL has a scheme we support.

```